### PR TITLE
feature: improve `useDerivedSwapInfo` hook

### PR DESCRIPTION
**Pull request by adamazad**
# Summary

Improves the performance of `useDerivedSwapInfo` hook by moving all computing logic to a `useEffect` hook.

### CODE REVIEW

File: src/state/swap/hooks.ts [hunk 1]
```diff
@@ -109,150 +97,212 @@ export function tryParseAmount(value?: string, currency?: Currency, chainId?: nu
...
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      relevantTokenBalances[0]?.raw.toString(),
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      relevantTokenBalances[1]?.raw.toString(),
+      allowedSlippage,
+      recipient,
+      isExactIn,
+    ],
+    `account,useMultihops,recipientLookupComputed,chainId,inputCurrency?.address,outputCurrency?.address,parsedAmountString,relevantTokenBalances[0]?.raw.toString(),relevantTokenBalances[1]?.raw.toString(),allowedSlippage,recipient,isExactIn`
   )
```
Comment (adamazad) [inline src/state/swap/hooks.ts:163]: This hook might be too much - we can remove it. However, it helps to identify what causes the excessive re-renders.

File: src/state/swap/hooks.ts [hunk 2]
```diff
@@ -109,150 +97,212 @@ export function tryParseAmount(value?: string, currency?: Currency, chainId?: nu
...
- * @param trade to check for the given address
- * @param checksummedAddress address to check in the pairs and tokens
- */
-function involvesAddress(trade: UniswapV2Trade, checksummedAddress: string): boolean {
-  return (
-    trade.route.path.some(token => token.address === checksummedAddress) ||
-    trade.route.pairs.some(pair => pair.liquidityToken.address === checksummedAddress)
-  )
-}
-
```
Comment (adamazad) [inline src/state/swap/hooks.ts:129]: I remove this bit because it's an incomplete feature. It should either

**Review by MilanVojnovic95 — APPROVED** [review]
Tested and looks ok to me 👍

**Comment by wixzi** [comment]
I will give another look tomorrow and approve. Nice work @adamazad  in cleaning this up 💯

File: src/state/swap/hooks.ts [hunk 3]
```diff
@@ -109,150 +97,212 @@ export function tryParseAmount(value?: string, currency?: Currency, chainId?: nu
...
   const relevantTokenBalances = useCurrencyBalances(account ?? undefined, [
     inputCurrency ?? undefined,
     outputCurrency ?? undefined,
   ])
-
-  const isExactIn: boolean = independentField === Field.INPUT
+  // Internal state
+  const [loading, setLoading] = useState<boolean>(false)
+  const [inputError, setInputError] = useState<number | undefined>()
+  // const [trade, setTrade] = useState<Trade>()
```
Comment (berteotti) [inline src/state/swap/hooks.ts:134]: This can be deleted, right?

File: src/state/swap/hooks.ts [hunk 4]
```diff
@@ -109,150 +97,212 @@ export function tryParseAmount(value?: string, currency?: Currency, chainId?: nu
...
+    // Require a valid input amount
+    if (!parsedAmount) {
+      setInputError(SWAP_INPUT_ERRORS.ENTER_AMOUNT)
+      return
+    }
+    // Require sufficient input balance
+    if (inputCurrencyBalance && JSBI.greaterThan(parsedAmount.raw, inputCurrencyBalance.raw as JSBI)) {
+      setInputError(SWAP_INPUT_ERRORS.INSUFFICIENT_FUNDS)
+      return
+    }
```
Comment (berteotti) [inline src/state/swap/hooks.ts:187]: Should we stop trade simulation even if one of these errors occurs? I think we can still get and show the best trades and then show some errors. Here is my suggestion 👇

```suggestion
    // Require a connected wallet
    if (!account) {
      setInputError(SWAP_INPUT_ERRORS.CONNECT_WALLET)
    }
    // Require two currencies to be selected
    if (!inputCurrency || !outputCurrency) {
      setInputError(SWAP_INPUT_ERRORS.SELECT_TOKEN)
      return
    }
    // Require a valid input amount
    if (!parsedAmount) {
      setInputError(SWAP_INPUT_ERRORS.ENTER_AMOUNT)
      return
    }
    // Require sufficient input balance
    if (inputCurrencyBalance && JSBI.greaterThan(parsedAmount.raw, inputCurrencyBalance.raw as JSBI)) {
      setInputError(SWAP_INPUT_ERRORS.INSUFFICIENT_FUNDS)
    }
```

**Review by berteotti — COMMENTED** [review]
Good work 🚀 Left some comments

File: src/state/swap/hooks.ts (hunk 3 above)
Comment (adamazad) [reply src/state/swap/hooks.ts:134]: Good catch!

File: src/state/swap/hooks.ts (hunk 4 above)
Comment (adamazad) [reply src/state/swap/hooks.ts:187]: The goal of this PR is to save resources; the Eco Router consumes a lot. hence, the Eco Router should be triggered when the right parameters are set. It's not rational for users to request a quote from the Eco Router when they don't have sufficient funds in the first place :)

File: src/state/swap/hooks.ts [hunk 5]
```diff
@@ -109,150 +97,212 @@ export function tryParseAmount(value?: string, currency?: Currency, chainId?: nu
...
   ])
-
-  const isExactIn: boolean = independentField === Field.INPUT
+  // Internal state
+  const [loading, setLoading] = useState<boolean>(false)
+  const [inputError, setInputError] = useState<number | undefined>()
+  // const [trade, setTrade] = useState<Trade>()
+  const [allPlatformTrades, setAllPlatformTrades] = useState<Trade[]>([])
+  // Computed on the fly state
+  const isExactIn = useMemo(() => independentField === Field.INPUT, [independentField])
```
Comment (Diogomartf) [inline src/state/swap/hooks.ts:137]: Shouldn't `useMemo` be used only on expensive/high resources function?

File: src/index.tsx [hunk 6]
```diff
@@ -19,6 +20,9 @@ import UserUpdater from './state/user/updater'
 import ThemeProvider, { FixedGlobalStyle, ThemedGlobalStyle } from './theme'
 import getLibrary from './utils/getLibrary'
 
+// Enables use of the useWhatChanged hook in dev environment
+setUseWhatChange(process.env.NODE_ENV === 'development')
```
Comment (Diogomartf) [inline src/index.tsx:24]: I'm always afraid that we make a typo on "development", should we make it a constant?

**Comment by wixzi** [comment]
https://user-images.githubusercontent.com/102727631/178599235-90f20253-2f61-433f-8d56-b0f9836a791d.mov

Something is missing 🐛 . 

Not a biggie but: The `@simbathesailor/use-what-changed` is logging in production build also.

**Comment by adamazad** [comment]
> Screen.Recording.2022-07-12.at.23.31.36.mov 
> Something is missing 🐛 .
> 
> Not a biggie but: The `@simbathesailor/use-what-changed` is logging in production build also.

> 

Different implementation than the README.md :/ 

https://github.com/simbathesailor/use-what-changed/blob/4fbf321cb18b122fa2a02c046e1bf3249f5836c5/src/useWhatChanged.tsx#L8-L11

fixed in [08f9843](https://github.com/levelkdev/swapr-dapp/pull/1227/commits/08f9843a8b06559263caa5e27b664f866add259e)

**Review by berteotti — APPROVED** [review]

**Pull request merged**